augur/dates.py

```diff
@@ -1,73 +1,271 @@
 import argparse
 import datetime
-from textwrap import dedent
 import isodate
 import pandas as pd
 import re
 import treetime.utils
+from textwrap import dedent
+from functools import lru_cache
+from typing import Any
+
+
+class InvalidDateFormat(ValueError):
+    pass
+
 
 from augur.util_support.date_disambiguator import DateDisambiguator
 
 SUPPORTED_DATE_HELP_TEXT = dedent("""\
     1. an Augur-style numeric date with the year as the integer part (e.g. 2020.42) or
     2. a date in ISO 8601 date format (i.e. YYYY-MM-DD) (e.g. '2020-06-04') or
-    3. a backwards-looking relative date in ISO 8601 duration format with optional P prefix (e.g. '1W', 'P1W')
+    3. an ambiguous date in ISO 8601-like format (e.g. '2020-06-XX', '2020-XX-XX') or
+    4. an incomplete date in ISO 8601-like format (e.g. '2020-06', '2020') or
+    5. a backwards-looking relative date in ISO 8601 duration format with optional P prefix (e.g. '1W', 'P1W')
 """)
 
-def numeric_date(date):
-    """
-    Converts the given *date* string to a :py:class:`float`.
-
-    *date* may be given as:
-    1. A string or float (number) with year as the integer part
-    2. A string in the YYYY-MM-DD (ISO 8601) syntax
-    3. A string representing a relative date (duration before datetime.date.today())
-
-    >>> numeric_date("2020.42")
-    2020.42
-    >>> numeric_date("2020-06-04")
-    2020.42486...
-    >>> import datetime, isodate, treetime
-    >>> numeric_date("1W") == treetime.utils.numeric_date(datetime.date.today() - isodate.parse_duration("P1W"))
-    True
+# float, negative ok
+# note: year-only is treated as incomplete ambiguous and must be non-negative (see RE_YEAR_ONLY)
+RE_NUMERIC_DATE = re.compile(r'^-*\d+\.\d+$')
+
+# complete ISO 8601 date
+# e.g. 2018-03-25
+RE_ISO_8601_DATE = re.compile(r'^\d{4}-\d{2}-\d{2}$')
+
+# complete ambiguous ISO 8601 date
+# e.g. 2018-03-XX
+RE_AMBIGUOUS_ISO_8601_DATE = re.compile(r'^[\dX]{4}-[\dX]{2}-[\dX]{2}$')
```

Comment:
Is this intended to allow values like `2022-01-X2` or `XX20-01-18`?

Reply:
Can't speak for the intention here, but we shouldn't allow `X` in a date if followed by a number.

Reply:
Thanks for catching this. The regex really needs some work!

I'm not sure what I'm intending here and what is desired. Created a separate issue where this should be clarified before continuing work here: #989

Reply:
The conclusion from #989 is to not support partial ambiguity here. 6a18b35c15b7892deef4a7baff3b445dab8a4a81 reflects that decision.